sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -1374,32 +1374,104 @@ case class Pivot(
   override protected def withNewChildInternal(newChild: LogicalPlan): Pivot = copy(child = newChild)
 }
 
+/**
+ * Expression for [[Unpivot]] for one unpivot value column (one or more expressions)
+ * and an optional alias. This node itself is not evaluable and resolvable.
+ * Only its children are to be resolved.
+ *
+ * @param exprs expressions to unpivot
+ * @param alias optional alias
+ */
+case class UnpivotExpr(exprs: Seq[NamedExpression], alias: Option[String]) extends Unevaluable {
```

Comment:
This reminds me of `ArraysZip`, where we need to enforce named expressions.

We can follow it to avoid analyzer hacks here and there
```
case class UnpivotExpr(exprs: Seq[Expression], names: Seq[Expression], alias: Option[String]) extends Unevaluable {
  def this(exprs: Seq[Expression], alias: Option[String]) = {
    this(exprs, exprs.map {
      case u: UnresolvedAttribute => Literal(u.nameParts.last)
      case e: NamedExpression if e.resolved => Literal(e.name)
      ...
    }, alias)
  }
}
```